compose: TestComposeCompare failed [no locations to init GEOS]

compose.TestComposeCompare [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_ComposeTestsBazel/15314240?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_ComposeTestsBazel/15314240?buildTab=artifacts#/) on release-23.2 @ [560157ccb797816924a12b123da0d3bf839551d1](https://github.com/cockroachdb/cockroach/commits/560157ccb797816924a12b123da0d3bf839551d1):


```
        cockroach2_1  | clusterID:           b293e98b-07e0-4892-b8c0-4c25fc5fd256
        cockroach2_1  | status:              initialized new cluster
        cockroach2_1  | nodeID:              1
        test_1        | === RUN   TestCompare/mutators
        test_1        |     compare_test.go:113: starting test: mutators
        test_1        | --- FAIL: TestCompare (1.80s)
        test_1        |     --- FAIL: TestCompare/mutators (0.00s)
        test_1        | panic: geos: no locations to init GEOS [recovered]
        test_1        | 	panic: geos: no locations to init GEOS
        test_1        | 
        test_1        | goroutine 109 [running]:
        test_1        | testing.tRunner.func1.2({0x2275120, 0xc00007e7a0})
        test_1        | 	GOROOT/src/testing/testing.go:1545 +0x238
        test_1        | testing.tRunner.func1()
        test_1        | 	GOROOT/src/testing/testing.go:1548 +0x397
        test_1        | panic({0x2275120?, 0xc00007e7a0?})
        test_1        | 	GOROOT/src/runtime/panic.go:914 +0x21f
        test_1        | github.com/cockroachdb/cockroach/pkg/sql/randgen.encodeInvertedIndexHistogramUpperBounds(0x46dc780?, {0x3212e78?, 0xc000a22ab0?})
        test_1        | 	github.com/cockroachdb/cockroach/pkg/sql/randgen/mutator.go:375 +0x436
        test_1        | github.com/cockroachdb/cockroach/pkg/sql/randgen.randHistogram(0x31d17c0?, 0x46dc780)
        test_1        | 	github.com/cockroachdb/cockroach/pkg/sql/randgen/mutator.go:308 +0x12c
        test_1        | github.com/cockroachdb/cockroach/pkg/sql/randgen.statisticsMutator.func1(0xc000714480)
        test_1        | 	github.com/cockroachdb/cockroach/pkg/sql/randgen/mutator.go:225 +0x7f
        test_1        | github.com/cockroachdb/cockroach/pkg/sql/randgen.statisticsMutator(0xc0006331d0, {0xc000633200?, 0x3, 0xc0006331d0?})
        test_1        | 	github.com/cockroachdb/cockroach/pkg/sql/randgen/mutator.go:258 +0x89d
        test_1        | github.com/cockroachdb/cockroach/pkg/sql/randgen.MultiStatementMutation.Mutate(0x20734a0?, 0x2478e12?, {0xc000633200?, 0x4a0f9c8?, 0x2?})
        test_1        | 	github.com/cockroachdb/cockroach/pkg/sql/randgen/mutator.go:96 +0x28
        test_1        | github.com/cockroachdb/cockroach/pkg/sql/randgen.RandCreateTables(0xc0007abf60?, {0x2478e12, 0x5}, 0x3, 0xa0?, {0xc001fd3a40, 0x3, 0x7fcf6ac41108?})
        test_1        | 	github.com/cockroachdb/cockroach/pkg/sql/randgen/schema.go:93 +0x131
        test_1        | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.randTablesN(0xc00205fcc8?, 0x1?, {0x0, 0x0})
        test_1        | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/setup.go:118 +0x15e
        test_1        | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.randTables(0xc000540a00?)
        test_1        | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/setup.go:98 +0x2f
        test_1        | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.init.wrapCommonSetup.func28(0x265fa28?)
        test_1        | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/setup.go:51 +0x1b
        test_1        | github.com/cockroachdb/cockroach/pkg/compose/compare/compare.TestCompare.func2(0xc0020d5860)
        test_1        | 	github.com/cockroachdb/cockroach/pkg/compose/compare/compare/compare_test.go:115 +0xe2
        test_1        | testing.tRunner(0xc0020d5860, 0xc000633140)
        test_1        | 	GOROOT/src/testing/testing.go:1595 +0xff
        test_1        | created by testing.(*T).Run in goroutine 7
        test_1        | 	GOROOT/src/testing/testing.go:1648 +0x3ad
        compare_test_1 exited with code 2
        Stopping compare_cockroach1_1 ... 
        Stopping compare_cockroach2_1 ... 
        Stopping compare_cockroach1_1 ... done
        Stopping compare_cockroach2_1 ... done
        Aborting on container exit...
        
    compose_test.go:143: exit status 2
--- FAIL: TestComposeCompare (9.04s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #124017 compose: TestComposeCompare failed [O-robot T-sql-foundations branch-release-24.1.0-rc]
- #123188 compose: TestComposeCompare failed [no locations to init GEOS] [O-robot T-sql-foundations branch-release-23.2.5-rc]
- #89361 compose: TestComposeCompare failed [O-robot T-sql-foundations branch-release-22.2]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestComposeCompare.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-38854

--- a/pkg/compose/BUILD.bazel
+++ b/pkg/compose/BUILD.bazel
@@ -14,6 +14,7 @@ go_test(
     ],
     deps = [
         "//pkg/build/bazel",
+        "//pkg/testutils/datapathutils",
         "//pkg/testutils/skip",
         "//pkg/util/envutil",
     ],

--- a/pkg/compose/compare/compare/BUILD.bazel
+++ b/pkg/compose/compare/compare/BUILD.bazel
@@ -10,6 +10,7 @@ go_test(
     visibility = ["//pkg/compose:__subpackages__"],
     deps = [
         "//pkg/cmd/cmpconn",
+        "//pkg/geo/geos",
         "//pkg/internal/sqlsmith",
         "//pkg/sql/randgen",
         "//pkg/testutils",

--- a/pkg/compose/compare/compare/compare_test.go
+++ b/pkg/compose/compare/compare/compare_test.go
@@ -15,11 +15,13 @@ import (
 	"flag"
 	"fmt"
 	"os"
+	"path"
 	"path/filepath"
 	"testing"
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/cmd/cmpconn"
+	"github.com/cockroachdb/cockroach/pkg/geo/geos"
 	"github.com/cockroachdb/cockroach/pkg/internal/sqlsmith"
 	"github.com/cockroachdb/cockroach/pkg/sql/randgen"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
@@ -42,6 +44,16 @@ func TestCompare(t *testing.T) {
 	// N.B. randomized SQL workload performed by this test may require CCL
 	var license = envutil.EnvOrDefaultString("COCKROACH_DEV_LICENSE", "")
 	require.NotEmptyf(t, license, "COCKROACH_DEV_LICENSE must be set")
+
+	// Initialize GEOS libraries so that the test can use geospatial types.
+	workingDir, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = geos.EnsureInit(geos.EnsureInitErrorDisplayPrivate, path.Join(workingDir, "lib"))
+	if err != nil {
+		t.Fatal(err)
+	}
 
 	uris := map[string]struct {
 		addr string

--- a/pkg/compose/compare/docker-compose.yml
+++ b/pkg/compose/compare/docker-compose.yml
@@ -3,20 +3,26 @@ services:
     # We use a docker image mirror to avoid pulling from 3rd party repos, which sometimes have reliability issues.
     # See https://cockroachlabs.atlassian.net/wiki/spaces/devinf/pages/3462594561/Docker+image+sync for the details.
     image: us-east1-docker.pkg.dev/crl-docker-sync/docker-mirror/docker.io/library/ubuntu:xenial-20170214
-    command: /cockroach/cockroach start-single-node --insecure --log-dir=/cockroach/logs --spatial-libs=/cockroach/lib --listen-addr cockroach1
+    user: ${UID}:${GID}
+    command: /cockroach/cockroach start-single-node --insecure --store=/cockroach/store --spatial-libs=/cockroach/lib --listen-addr cockroach1
     volumes:
       - "${COCKROACH_PATH}:/cockroach/cockroach"
       - "${LIBGEOS_DIR_PATH}:/cockroach/lib"
-      - "${COMPARE_DIR_PATH}/logs1:/cockroach/logs"
+      - "${COMPARE_DIR_PATH}/store1:/cockroach/store"
+      - /etc/passwd:/etc/passwd:ro
+      - /etc/group:/etc/group:ro
   cockroach2:
     # We use a docker image mirror to avoid pulling from 3rd party repos, which sometimes have reliability issues.
     # See https://cockroachlabs.atlassian.net/wiki/spaces/devinf/pages/3462594561/Docker+image+sync for the details.
     image: us-east1-docker.pkg.dev/crl-docker-sync/docker-mirror/docker.io/library/ubuntu:xenial-20170214
-    command: /cockroach/cockroach start-single-node --insecure --log-dir=/cockroach/logs --spatial-libs=/cockroach/lib --listen-addr cockroach2
+    user: ${UID}:${GID}
+    command: /cockroach/cockroach start-single-node --insecure --store=/cockroach/store --spatial-libs=/cockroach/lib --listen-addr cockroach2
     volumes:
       - "${COCKROACH_PATH}:/cockroach/cockroach"
       - "${LIBGEOS_DIR_PATH}:/cockroach/lib"
-      - "${COMPARE_DIR_PATH}/logs2:/cockroach/logs"
+      - "${COMPARE_DIR_PATH}/store2:/cockroach/store"
+      - /etc/passwd:/etc/passwd:ro
+      - /etc/group:/etc/group:ro
   test:
     # We use a docker image mirror to avoid pulling from 3rd party repos, which sometimes have reliability issues.
     # See https://cockroachlabs.atlassian.net/wiki/spaces/devinf/pages/3462594561/Docker+image+sync for the details.

--- a/pkg/compose/compare/docker-compose.yml
+++ b/pkg/compose/compare/docker-compose.yml
@@ -3,18 +3,20 @@ services:
     # We use a docker image mirror to avoid pulling from 3rd party repos, which sometimes have reliability issues.
     # See https://cockroachlabs.atlassian.net/wiki/spaces/devinf/pages/3462594561/Docker+image+sync for the details.
     image: us-east1-docker.pkg.dev/crl-docker-sync/docker-mirror/docker.io/library/ubuntu:xenial-20170214
-    command: /cockroach/cockroach start-single-node --insecure --listen-addr cockroach1
+    command: /cockroach/cockroach start-single-node --insecure --log-dir=/cockroach/logs --spatial-libs=/cockroach/lib --listen-addr cockroach1
     volumes:
       - "${COCKROACH_PATH}:/cockroach/cockroach"
       - "${LIBGEOS_DIR_PATH}:/cockroach/lib"
+      - "${COMPARE_DIR_PATH}/logs1:/cockroach/logs"
   cockroach2:
     # We use a docker image mirror to avoid pulling from 3rd party repos, which sometimes have reliability issues.
     # See https://cockroachlabs.atlassian.net/wiki/spaces/devinf/pages/3462594561/Docker+image+sync for the details.
     image: us-east1-docker.pkg.dev/crl-docker-sync/docker-mirror/docker.io/library/ubuntu:xenial-20170214
-    command: /cockroach/cockroach start-single-node --insecure --listen-addr cockroach2
+    command: /cockroach/cockroach start-single-node --insecure --log-dir=/cockroach/logs --spatial-libs=/cockroach/lib --listen-addr cockroach2
     volumes:
       - "${COCKROACH_PATH}:/cockroach/cockroach"
       - "${LIBGEOS_DIR_PATH}:/cockroach/lib"
+      - "${COMPARE_DIR_PATH}/logs2:/cockroach/logs"
   test:
     # We use a docker image mirror to avoid pulling from 3rd party repos, which sometimes have reliability issues.
     # See https://cockroachlabs.atlassian.net/wiki/spaces/devinf/pages/3462594561/Docker+image+sync for the details.

--- a/pkg/compose/compose_test.go
+++ b/pkg/compose/compose_test.go
@@ -22,6 +22,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/build/bazel"
+	"github.com/cockroachdb/cockroach/pkg/testutils/datapathutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/envutil"
 )
@@ -76,10 +77,18 @@ func TestComposeCompare(t *testing.T) {
 		// start up docker-compose, but the files themselves will be
 		// Bazel-built symlinks. We want to copy these files to a
 		// different temporary location.
-		composeBinsDir := t.TempDir()
-		compareDir = composeBinsDir
-		cockroachBin = filepath.Join(composeBinsDir, "cockroach")
-		libGeosDir = filepath.Join(composeBinsDir, "lib")
+		compareDir, err = os.MkdirTemp(datapathutils.DebuggableTempDir(), "TestComposeCompare")
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Cleanup(func() {
+			if t.Failed() {
+				return
+			}
+			_ = os.RemoveAll(compareDir)
+		})
+		cockroachBin = filepath.Join(compareDir, "cockroach")
+		libGeosDir = filepath.Join(compareDir, "lib")
 		if err = os.MkdirAll(libGeosDir, 0755); err != nil {
 			t.Fatal(err)
 		}
@@ -93,7 +102,7 @@ func TestComposeCompare(t *testing.T) {
 				t.Fatal(err)
 			}
 		}
-		if err = copyBin(*flagCompare, filepath.Join(composeBinsDir, "compare.test")); err != nil {
+		if err = copyBin(*flagCompare, filepath.Join(compareDir, "compare.test")); err != nil {
 			t.Fatal(err)
 		}
 		if *flagArtifacts == "" {
@@ -137,6 +146,7 @@ func TestComposeCompare(t *testing.T) {
 		fmt.Sprintf("PATH=%s", os.Getenv("PATH")),
 		"COCKROACH_RUN_COMPOSE_COMPARE=true",
 	}
+	t.Logf("running: %s", cmd)
 	out, err := cmd.CombinedOutput()
 	t.Log(string(out))
 	if err != nil {

--- a/pkg/compose/compose_test.go
+++ b/pkg/compose/compose_test.go
@@ -17,6 +17,7 @@ import (
 	"io"
 	"os"
 	"os/exec"
+	"os/user"
 	"path/filepath"
 	"testing"
 	"time"
@@ -87,6 +88,12 @@ func TestComposeCompare(t *testing.T) {
 			}
 			_ = os.RemoveAll(compareDir)
 		})
+		if err = os.MkdirAll(filepath.Join(compareDir, "store1"), 0755); err != nil {
+			t.Fatal(err)
+		}
+		if err = os.MkdirAll(filepath.Join(compareDir, "store2"), 0755); err != nil {
+			t.Fatal(err)
+		}
 		cockroachBin = filepath.Join(compareDir, "cockroach")
 		libGeosDir = filepath.Join(compareDir, "lib")
 		if err = os.MkdirAll(libGeosDir, 0755); err != nil {
@@ -135,7 +142,13 @@ func TestComposeCompare(t *testing.T) {
 		"--force-recreate",
 		"--exit-code-from", "test",
 	)
+	userInfo, err := user.Current()
+	if err != nil {
+		t.Fatal(err)
+	}
 	cmd.Env = []string{
+		fmt.Sprintf("UID=%s", userInfo.Uid),
+		fmt.Sprintf("GID=%s", userInfo.Gid),
 		fmt.Sprintf("EACH=%s", *flagEach),
 		fmt.Sprintf("TESTS=%s", *flagTests),
 		fmt.Sprintf("COCKROACH_PATH=%s", cockroachBin),

